### PR TITLE
feat: `metadata.component` for poetry 

### PR DESCRIPTION
**Pull request by italvi**
This PR tries to address #391 

Currently I only check for a pyproject.toml. As this will be used in combination with poetry I only added this feature there.

### CODE REVIEW

File: cyclonedx_py/client.py [hunk 1]
```diff
@@ -111,7 +111,7 @@ def get_output(self) -> BaseOutput:
                   '',
                   sep='\n', file=sys.stderr)
 
-        bom = Bom(components=filter(self._component_filter, parser.get_components()))
+        bom = Bom(components=filter(self._component_filter, parser.get_components()), metadata=parser._metadata)
```
Comment (jkowalleck) [inline cyclonedx_py/client.py:114]: :x: do not access protected/private properties! 
use the public getters/properties!

File: cyclonedx_py/parser/poetry.py [hunk 2]
```diff
@@ -97,19 +99,29 @@ def __init__(
                     del error
             self._components.append(component)
 
+        if pyproject_toml_contents:
+            poetry_toml = load_toml(pyproject_toml_contents)
+            poetry_toml_metadata = poetry_toml.get('tool', '').get('poetry', '')
```
Comment (jkowalleck) [inline cyclonedx_py/parser/poetry.py:104]: wrong types for defaults
```patch
-             poetry_toml_metadata = poetry_toml.get('tool', '').get('poetry', '')
+             poetry_toml_metadata = poetry_toml.get('tool', {}).get('poetry', '')
```

**Review by jkowalleck — CHANGES_REQUESTED** [review]

**Comment by jkowalleck** [comment]
converted this PR to "draft" as the target branch and source branch are not the desired ones. 
see https://github.com/CycloneDX/cyclonedx-python/pull/542#issuecomment-1517608658

**Comment by italvi** [comment]
> Hello @italvi , thank you for the contribution.
> 
> This project ins currently transitioning to v4. This means: the current develop-branch is [`dev/4.x.x`](https://github.com/CycloneDX/cyclonedx-python/tree/dev/4.x.x)
> 
> Could you develop your feature from there and pullrequest it to that branch?

@jkowalleck could you please further elaborate? I forked the project and then made my changes on the branch dev/4.x.x there

File: cyclonedx_py/client.py (hunk 1 above)
Comment (italvi) [reply cyclonedx_py/client.py:114]: The current implementation of the `BaseParser` does not provide a get_metadata. Hence the direct access to the properties

Comment (jkowalleck) [reply cyclonedx_py/client.py:114]: feel free to create one.

Comment (italvi) [reply cyclonedx_py/client.py:114]: would you prefer to extend it in the python-lib or shall I write a ExtendedBaseParser in this project?

**Comment by jkowalleck** [comment]
superseded by https://github.com/CycloneDX/cyclonedx-python/pull/605

**Pull request closed without merging**